app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt

```diff
@@ -49,17 +50,126 @@ class WallpaperFileManager(
+     * @param downloadThumbnail Suspend function used to download the thumbnail for a wallpaper
+     * available remote.
+     * @return List of migrated wallpapers.
+     */
+    suspend fun migrateAndCleanLegacyWallpapers(
+        remoteAvailableWallpapers: List<Wallpaper>,
+        localAvailableWallpapers: List<Wallpaper>,
+        downloadThumbnail: suspend (Wallpaper) -> Wallpaper.ImageFileState,
+    ): List<Wallpaper> = withContext(Dispatchers.IO) {
+        val legacyPortraitDirectory = File(storageRootDirectory, "wallpapers/portrait/light")
+        val legacyLandscapeDirectory = File(storageRootDirectory, "wallpapers/landscape/light")
+
+        val legacyWallpapers = legacyPortraitDirectory.listFiles()
+            ?.filter { file ->
+                file.isFile && !localAvailableWallpapers.map { it.name }
+                    .contains(file.nameWithoutExtension)
+            }
+            ?.mapNotNull { portraitFile ->
+                val landscapeFile = File(legacyLandscapeDirectory, portraitFile.name)
+                if (landscapeFile.exists()) {
+                    Pair(portraitFile, landscapeFile)
+                } else {
+                    null
+                }
+            }
+
+        val migratedWallpapers = legacyWallpapers?.map { pair ->
+            getWallpaperByName(pair.first.nameWithoutExtension, remoteAvailableWallpapers)
+        }
+
+        legacyWallpapers?.let {
+            migrateLegacyWallpapers(remoteAvailableWallpapers, it, downloadThumbnail)
+        }
+        migratedWallpapers ?: listOf()
+    }
+
+    private fun getWallpaperByName(name: String, remoteWallpapers: List<Wallpaper>): Wallpaper =
+        remoteWallpapers.firstOrNull { it.name == name } ?: Wallpaper(
+            name = name,
+            collection = Wallpaper.DefaultCollection,
+            textColor = null,
+            cardColor = null,
+            thumbnailFileState = Wallpaper.ImageFileState.Downloaded,
+            assetsFileState = Wallpaper.ImageFileState.Downloaded,
+        )
+
+    @Suppress("TooGenericExceptionCaught")
+    private fun migrateLegacyWallpapers(
+        availableWallpapers: List<Wallpaper>,
+        legacyWallpapers: List<Pair<File, File>>,
+        downloadThumbnail: suspend (Wallpaper) -> Wallpaper.ImageFileState,
+    ) = scope.launch {
+        legacyWallpapers.forEach { pair ->
+            val portraitFile = pair.first
+            val landscapeFile = pair.second
+            val name = portraitFile.nameWithoutExtension
+            val directory = "wallpapers/${name.lowercase()}"
+            val remoteWallpaper = availableWallpapers.firstOrNull { it.name == name }
+
+            try {
+                // If there is a thumbnail available for the current wallpaper,
+                // attempt to download it
+                if (remoteWallpaper == null ||
+                    downloadThumbnail(remoteWallpaper) == Wallpaper.ImageFileState.Error
```

Comment:
I applaud the thoroughness of downloading a correct thumbnail if it exists. In the interest of simplicity, I think it would also be fine to just use the portrait image as the thumbnail. That would also ensure the thumbnail matches the legacy version of the wallpaper.